torch/optim/nadam.py

```diff
@@ -299,24 +308,38 @@ def _multi_tensor_nadam(params: List[Tensor],
                         momentum_decay: float,
                         eps: float,
                         decoupled_weight_decay: bool,
+                        capturable: bool,
                         differentiable: bool):
 
     if len(params) == 0:
         return
 
     assert not differentiable, "_foreach ops don't support autograd"
 
-    grouped_tensors = Optimizer._group_tensors_by_device_and_dtype([params, grads, exp_avgs, exp_avg_sqs, mu_products, state_steps])
+    grouped_tensors = Optimizer._group_tensors_by_device_and_dtype(
+        [params, grads, exp_avgs, exp_avg_sqs, mu_products, state_steps])
     for ((grouped_params, grouped_grads, grouped_exp_avgs,
          grouped_exp_avg_sqs, grouped_mu_products, grouped_state_steps), _) in grouped_tensors.values():
 
         # update steps
         torch._foreach_add_(grouped_state_steps, 1)
 
-        bias_correction2 = [1 - beta2 ** _get_value(step) for step in grouped_state_steps]
-        mus = [beta1 * (1. - 0.5 * (0.96 ** (_get_value(step) * momentum_decay))) for step in grouped_state_steps]
-        mu_nexts = [beta1 * (1. - 0.5 * (0.96 ** ((_get_value(step) + 1) * momentum_decay)))
-                    for step in grouped_state_steps]
+        # scalar python math is faster in eager
+        # however, it is not visible to cudagraphs
+        # and can be slow to trace
+        if capturable:
+            bias_correction2 = torch._foreach_pow(beta2, grouped_state_steps)
+            bias_correction2 = torch._foreach_add(torch._foreach_mul(bias_correction2, -1.0), 1.0)
```

Comment:
All of the nested foreach ops here increase peak memory usage. It would be good to have one op per line and to use the inplace ops to minimize peak memory usage.